smithy-lang/smithy-ruby
# Add Telemetry support

## Description
Observability is the extent to which a system's current state can be inferred from the data it emits. The data emitted is commonly referred to as telemetry. This PR will set up the groundwork for SDK instrumentation and current/future telemetry providers. 

In addition, we will support OpenTelemetry as a telemetry provider. 

**Related PR**: https://github.com/alextwoods/aws-sdk-ruby-v4/pull/54

## Changes
* Add `opentelemetry-sdk` gem for testing purposes
* Codegen
  * Update FakeProtocol's BuilderGenerator to render `http_method` (intended for span attribute testing)
  * Update `config_spec.rb` to include the new Config, `telemetry_provider`.
  * Add `telemetry_spec.rb` to test Config and Client behaviors (includes OTel-related testings)
  * Add  smithy file to test isolate and test telemetry components
  * Update ClientGenerator to include render telemetry code for each operation
  * Update MiddlewareBuilder to include new config, `telemetry_provider`
  * Update RubySettings to generate/access `SDK_ID` (if no SDK ID is present, we use Service Shape name)
* Hearth
  * Update Context to have access to a tracer
  * Update Send Middleware to emit telemetry (for both stub and send responses)
  * Add telemetry components, no-op classes and otel support (only tracing supported)
  * Add RBS signatures for telemetry components
  * Update `send_spec.rb` as adding telemetry support has changed how the tests are run
  * Add specs for telemetry components


~~* Add V4 support (separate PR in V4 repo)~~
~~* Create child span on a middleware level~~
~~* Update specs based on the child span + add more specs~~
~~* Add documentation~~

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
